modules/agar/src/test/ts/browser/ApproxStructureTest.ts

```diff
@@ -1,139 +1,426 @@
-
-  const struct1 = ApproxStructure.build((s, str, arr) =>
-    s.either([
-      s.element('span', {
-        classes: [
-          arr.has('hello'),
-          arr.not('fizzbuzz')
-        ]
-      }),
-      s.element('div', {
-        classes: [
-          arr.has('fizzbuzz'),
-          arr.not('hello')
-        ]
-      })
-    ]));
-
-  check(struct1, '<span class="hello"></span>');
-  check(struct1, '<div class="fizzbuzz"></span>');
-
-  const struct2 = ApproxStructure.build((s, str, arr) =>
-    s.element('div', {
-      children: [
-        s.oneOrMore(s.element('span', {
+            'selected': str.is('double'),
+            'data-key': str.contains('test')
+          },
           classes: [
-            arr.has('hello')
+            'test1', 'root'
+          ],
+          styles: {
+            display: str.is('block')
+          },
+          children: [
+            s.elementWithSpecifics('div', {
+              attrs: {
+                selected: str.is('true')
+              },
+              children: [
+                s.elementWithSpecifics('span', {
+                  attrs: {
+                    'data-ephox-id': str.startsWith('bl')
+                  },
+                  classes: [ 'disabled' ],
+                  html: str.is('span')
+                })
+              ]
+            }),
+            s.text(
+              str.is('words')
+            ),
+            s.anything()
+          ]
+        })), html);
+    });
+
+    it('TINY-9102: ApproxStructure either', () => {
+      const struct1 = ApproxStructure.build((s, _str, _arr) =>
+        s.either([
+          s.elementWithSpecifics('span', {
+            classes: [
+              'hello'
+            ]
+          }),
+          s.elementWithSpecifics('div', {
+            classes: [
+              'fizzbuzz'
+            ]
+          }),
+        ]));
+
+      check(struct1, '<span class="hello"></span>');
+      check(struct1, '<div class="fizzbuzz"></span>');
+    });
+
+    it('TINY-9102: ApproxStructure oneOrMore and zeroOrOne', () => {
+      const struct2 = ApproxStructure.build((s, _str, _arr) =>
+        s.elementWithSpecifics('div', {
+          children: [
+            s.oneOrMore(s.elementWithSpecifics('span', {
+              classes: [
+                'hello'
+              ]
+            })),
+            s.elementWithSpecifics('div', {}),
+            s.zeroOrOne(s.elementWithSpecifics('span', {
+              classes: [
+                'bye'
+              ]
+            }))
+          ]
+        }));
+
+      check(struct2, '<div><span class="hello"></span><div></div></div>');
+      check(struct2, '<div><span class="hello"></span><div></div><span class="bye"></span></div>');
+      check(struct2, '<div><span class="hello"></span><span class="hello"></span><span class="hello"></span><span class="hello"></span><div></div><span class="bye"></span></div>');
+    });
+
+    it('TINY-9102: ApproxStructure text combineSiblings', () => {
```

Comment:
Not sure what combineSiblings means here

Reply:
Looks like it means that adjacent text nodes can be normalised into one in the expected.